pkg/restore/split_client.go

```diff
@@ -0,0 +1,353 @@
+	// SetPlacementRule insert or update a placement rule to PD.
+	SetPlacementRule(ctx context.Context, rule placement.Rule) error
+	// DeletePlacementRule removes a placement rule from PD.
+	DeletePlacementRule(ctx context.Context, groupID, ruleID string) error
+	// SetStoreLabel add or update specified label of stores. If labelValue
+	// is empty, it clears the label.
+	SetStoresLabel(ctx context.Context, stores []uint64, labelKey, labelValue string) error
+}
+
+// pdClient is a wrapper of pd client, can be used by RegionSplitter.
+type pdClient struct {
+	mu         sync.Mutex
+	client     pd.Client
+	storeCache map[uint64]*metapb.Store
+}
+
+// NewSplitClient returns a client used by RegionSplitter.
+func NewSplitClient(client pd.Client) SplitClient {
+	return &pdClient{
+		client:     client,
+		storeCache: make(map[uint64]*metapb.Store),
+	}
+}
+
+func (c *pdClient) GetStore(ctx context.Context, storeID uint64) (*metapb.Store, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	store, ok := c.storeCache[storeID]
+	if ok {
+		return store, nil
+	}
+	store, err := c.client.GetStore(ctx, storeID)
+	if err != nil {
+		return nil, err
+	}
+	c.storeCache[storeID] = store
+	return store, nil
+
+}
+
+func (c *pdClient) GetRegion(ctx context.Context, key []byte) (*RegionInfo, error) {
+	region, leader, err := c.client.GetRegion(ctx, key)
+	if err != nil {
+		return nil, err
+	}
+	if region == nil {
+		return nil, nil
+	}
+	return &RegionInfo{
+		Region: region,
+		Leader: leader,
+	}, nil
+}
+
+func (c *pdClient) GetRegionByID(ctx context.Context, regionID uint64) (*RegionInfo, error) {
+	region, leader, err := c.client.GetRegionByID(ctx, regionID)
+	if err != nil {
+		return nil, err
+	}
+	if region == nil {
+		return nil, nil
+	}
+	return &RegionInfo{
+		Region: region,
+		Leader: leader,
+	}, nil
+}
+
+func (c *pdClient) SplitRegion(ctx context.Context, regionInfo *RegionInfo, key []byte) (*RegionInfo, error) {
+	var peer *metapb.Peer
+	if regionInfo.Leader != nil {
+		peer = regionInfo.Leader
+	} else {
+		if len(regionInfo.Region.Peers) == 0 {
+			return nil, errors.New("region does not have peer")
+		}
+		peer = regionInfo.Region.Peers[0]
+	}
+	storeID := peer.GetStoreId()
+	store, err := c.GetStore(ctx, storeID)
+	if err != nil {
+		return nil, err
+	}
+	conn, err := grpc.Dial(store.GetAddress(), grpc.WithInsecure())
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	client := tikvpb.NewTikvClient(conn)
+	resp, err := client.SplitRegion(ctx, &kvrpcpb.SplitRegionRequest{
+		Context: &kvrpcpb.Context{
+			RegionId:    regionInfo.Region.Id,
+			RegionEpoch: regionInfo.Region.RegionEpoch,
+			Peer:        peer,
+		},
+		SplitKey: key,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if resp.RegionError != nil {
+		return nil, errors.Errorf("split region failed: region=%v, key=%x, err=%v", regionInfo.Region, key, resp.RegionError)
+	}
+
+	// BUG: Left is deprecated, it may be nil even if split is succeed!
```

Comment:
PTAL @5kbpers

Reply:
It would be found in the `Regions` if resp (see L156~L164)